crates/ruff_linter/src/rules/ruff/rules/parenthesize_logical_operators.rs

```diff
@@ -75,18 +83,30 @@ pub(crate) fn parenthesize_chained_logical_operators(
                     ..
                 },
             ) => {
+                let locator = checker.locator();
+                let source_range = bool_op.range();
                 if parenthesized_range(
                     bool_op.into(),
                     expr.into(),
                     checker.indexer().comment_ranges(),
-                    checker.locator().contents(),
+                    locator.contents(),
                 )
                 .is_none()
                 {
-                    checker.diagnostics.push(Diagnostic::new(
-                        ParenthesizeChainedOperators,
-                        bool_op.range(),
-                    ));
+                    let mut diagnostic =
+                        Diagnostic::new(ParenthesizeChainedOperators, source_range);
+                    let line_length: u16 = checker.settings.line_length.into();
+                    if !locator.contains_line_break(source_range)
```

Comment:
Here's a nightmare scenario in the mypy codebase: https://github.com/python/mypy/blob/35f402c74205d373b81076ea82f507eb85161a25/mypy/messages.py#L2940-L2955

<details>

```py
def get_bad_protocol_flags():
    for name, subflags, superflags in all_flags:
        if (
            IS_CLASSVAR in subflags
            and IS_CLASSVAR not in superflags
            and IS_SETTABLE in superflags
            or IS_CLASSVAR in superflags
            and IS_CLASSVAR not in subflags
            or IS_SETTABLE in superflags
            and IS_SETTABLE not in subflags
            or IS_CLASS_OR_STATIC in superflags
            and IS_CLASS_OR_STATIC not in subflags
            or class_obj
            and IS_VAR in superflags
            and IS_CLASSVAR not in subflags
            or class_obj
            and IS_CLASSVAR in superflags
        ):
            bad_flags.append((name, subflags, superflags))
```

</details>

The naivest fix would be to do something like this, but it would be code that no autoformatter would accept:

<details>

```py
def get_bad_protocol_flags():
    for name, subflags, superflags in all_flags:
        if (
            (IS_CLASSVAR in subflags
            and IS_CLASSVAR not in superflags
            and IS_SETTABLE in superflags)
            or (IS_CLASSVAR in superflags
            and IS_CLASSVAR not in subflags)
            or (IS_SETTABLE in superflags
            and IS_SETTABLE not in subflags)
            or (IS_CLASS_OR_STATIC in superflags
            and IS_CLASS_OR_STATIC not in subflags)
            or (class_obj
            and IS_VAR in superflags
            and IS_CLASSVAR not in subflags)
            or (class_obj
            and IS_CLASSVAR in superflags)
        ):
            bad_flags.append((name, subflags, superflags))
```

</details>

A slightly less naive fix would be to do something like this:

<details>

```py
def get_bad_protocol_flags():
    for name, subflags, superflags in all_flags:
        if (
            (
                IS_CLASSVAR in subflags
                and IS_CLASSVAR not in superflags
                and IS_SETTABLE in superflags
            ) or (
                IS_CLASSVAR in superflags
                and IS_CLASSVAR not in subflags
            ) or (
                IS_SETTABLE in superflags
                and IS_SETTABLE not in subflags
            ) or (
                IS_CLASS_OR_STATIC in superflags
                and IS_CLASS_OR_STATIC not in subflags
            ) or (
                class_obj
                and IS_VAR in superflags
                and IS_CLASSVAR not in subflags
            ) or (
                class_obj
                and IS_CLASSVAR in superflags
            )
        ):
            bad_flags.append((name, subflags, superflags))
```

</details>

But what we _actually_ want is something like this:

<details>

```py
def get_bad_protocol_flags():
    for name, subflags, superflags in all_flags:
        if (
            (
                IS_CLASSVAR in subflags
                and IS_CLASSVAR not in superflags
                and IS_SETTABLE in superflags
            )
            or (IS_CLASSVAR in superflags and IS_CLASSVAR not in subflags)
            or (IS_SETTABLE in superflags and IS_SETTABLE not in subflags)
            or (IS_CLASS_OR_STATIC in superflags and IS_CLASS_OR_STATIC not in subflags)
            or (class_obj and IS_VAR in superflags and IS_CLASSVAR not in subflags)
            or (class_obj and IS_CLASSVAR in superflags)
         ):
            bad_flags.append((name, subflags, superflags))
```

</details>

I'm sure it's not impossible to implement it, but is it _worth_ the complexity? I'm not sure. It would also be even harder to get right if there were comments between the lines.

Reply:
@charliermarsh probably knows best of what we did in these cases in the past. I'm not to worried about the formatting of the fixed code. Most users use a code formatter that will make your fix look pretty ;) We also consider integrating the formatter into check, so that files get automatically formatted after the fixes were applied.

Reply:
It looks like the formatter actually _does_ do:

```python
def get_bad_protocol_flags():
    for name, subflags, superflags in all_flags:
        if (
            (
                IS_CLASSVAR in subflags
                and IS_CLASSVAR not in superflags
                and IS_SETTABLE in superflags
            )
            or (IS_CLASSVAR in superflags and IS_CLASSVAR not in subflags)
            or (IS_SETTABLE in superflags and IS_SETTABLE not in subflags)
            or (IS_CLASS_OR_STATIC in superflags and IS_CLASS_OR_STATIC not in subflags)
            or (class_obj and IS_VAR in superflags and IS_CLASSVAR not in subflags)
            or (class_obj and IS_CLASSVAR in superflags)
        ):
            bad_flags.append((name, subflags, superflags))
```

Which is nice.

I tend towards allowing the fix, since if we don't want it fixed, users should either turn off the rule _or_ we shouldn't be flagging it in these cases (somehow).

Reply:
Should I go for the naivest fix, or the "slightly less naive fix", do you think? I don't think I can do the "one we _actually_ want" without reimplementing the formatter in the linter 😄

Reply:
Nice and easy ^_^